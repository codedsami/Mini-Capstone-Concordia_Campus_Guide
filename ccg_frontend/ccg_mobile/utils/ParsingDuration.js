export const formatDuration = (seconds) => {
  if (seconds <= 0) return "Now";

  const days = calculateDays(seconds);
  seconds = remainingSecondsAfterDays(seconds);

  const hours = calculateHours(seconds);
  seconds = remainingSecondsAfterHours(seconds);

  const minutes = calculateMinutes(seconds);

  const parts = [];
  addTimePart(parts, days, "day");
  addTimePart(parts, hours, "hour");
  addTimePart(parts, minutes, "minute");

  return parts.join(", ");
};

// Extracted methods for calculations
export const calculateDays = (seconds) => Math.floor(seconds / 86400);
export const remainingSecondsAfterDays = (seconds) => seconds % 86400;

export const calculateHours = (seconds) => Math.floor(seconds / 3600);
export const remainingSecondsAfterHours = (seconds) => seconds % 3600;

export const calculateMinutes = (seconds) => Math.floor(seconds / 60);

// Extracted method for adding parts
export const addTimePart = (parts, value, unit) => {
  if (value > 0) {
    parts.push(`${value} ${unit}${value !== 1 ? "s" : ""}`);
  }
};
