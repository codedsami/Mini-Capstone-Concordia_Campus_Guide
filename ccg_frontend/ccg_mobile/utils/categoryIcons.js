// React Native doesn't allow us to use dynamic paths with require(), so we must use this logic.
const categoryIcons = {
  atm: require("../assets/poi_icons/atm.png"),
  bank: require("../assets/poi_icons/bank.png"),
  bar: require("../assets/poi_icons/bar.png"),
  bicycle_parking: require("../assets/poi_icons/bicycle_parking.png"),
  bicycle_rental: require("../assets/poi_icons/bicycle_rental.png"),
  bureau_de_change: require("../assets/poi_icons/bureau_de_change.png"),
  cafe: require("../assets/poi_icons/cafe.png"),
  car_rental: require("../assets/poi_icons/car_rental.png"),
  charging_station: require("../assets/poi_icons/charging_station.png"),
  childcare: require("../assets/poi_icons/childcare.png"),
  cinema: require("../assets/poi_icons/cinema.png"),
  clinic: require("../assets/poi_icons/clinic.png"),
  community_centre: require("../assets/poi_icons/community_centre.png"),
  convenience: require("../assets/poi_icons/convenience.png"),
  dentist: require("../assets/poi_icons/dentist.png"),
  doctors: require("../assets/poi_icons/doctors.png"),
  dojo: require("../assets/poi_icons/dojo.png"),
  driving_school: require("../assets/poi_icons/driving_school.png"),
  events_venue: require("../assets/poi_icons/events_venue.png"),
  fast_food: require("../assets/poi_icons/fast_food.png"),
  fuel: require("../assets/poi_icons/fuel.png"),
  ice_cream: require("../assets/poi_icons/ice_cream.png"),
  library: require("../assets/poi_icons/library.png"),
  nightclub: require("../assets/poi_icons/nightclub.png"),
  other: require("../assets/poi_icons/other.png"),
  parking_entrance: require("../assets/poi_icons/parking_entrance.png"),
  pharmacy: require("../assets/poi_icons/pharmacy.png"),
  police: require("../assets/poi_icons/police.png"),
  post_box: require("../assets/poi_icons/post_box.png"),
  post_office: require("../assets/poi_icons/post_office.png"),
  pub: require("../assets/poi_icons/pub.png"),
  restaurant: require("../assets/poi_icons/restaurant.png"),
  school: require("../assets/poi_icons/school.png"),
  social_facility: require("../assets/poi_icons/social_facility.png"),
  theatre: require("../assets/poi_icons/theatre.png"),
  townhall: require("../assets/poi_icons/townhall.png"),
  university: require("../assets/poi_icons/university.png"),
  waste_basket: require("../assets/poi_icons/waste_basket.png"),
};

export const getCategoryIcon = (category) => {
  return categoryIcons[category] || categoryIcons.university;
};

export const getCategoryIcons = () => categoryIcons;
