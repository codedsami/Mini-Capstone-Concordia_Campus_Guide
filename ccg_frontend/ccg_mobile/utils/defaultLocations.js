import locationService from "../services/LocationService";

export const getMyCurrentLocation = async () => {
  try {
    await locationService.startTrackingLocation();
    const currentLocation = locationService.getCurrentLocation();

    if (!currentLocation?.coords) {
      return getDefaultDestination();
    }

    return {
      location: {
        latitude: currentLocation.coords.latitude,
        longitude: currentLocation.coords.longitude,
      },
      civic_address: "Current Location",
    };
  } catch (error) {
    // Return a fallback location if unable to get current location
    return getDefaultDestination();
  }
};

export const getDefaultDestination = () => {
  return {
    location: {
      latitude: 45.4972159,
      longitude: -73.5794777,
    },
    civic_address: "Concordia University",
  };
};
