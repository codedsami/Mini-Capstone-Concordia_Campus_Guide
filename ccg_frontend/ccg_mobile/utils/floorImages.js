const floorImages = {
  H8: require("../assets/floors/Hall-8.png"),
  H9: require("../assets/floors/Hall-9.png"),
  H2: require("../assets/floors/H2.png"),
  H1: require("../assets/floors/H1.png"),
  MB1: require("../assets/floors/MB-1.png"),
  CC1: require("../assets/floors/CC1.png"),
};

export const getFloorImage = floor => {
  return floorImages[floor] || floorImages.H8;
};

export const getFloorImages = () => floorImages;
