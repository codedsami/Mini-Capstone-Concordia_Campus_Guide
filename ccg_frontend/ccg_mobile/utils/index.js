export const formatDuration = (seconds) => {
    if (seconds <= 0) return 'Now';
    const days = Math.floor(seconds / 86400);
    seconds %= 86400;
    const hours = Math.floor(seconds / 3600);
    seconds %= 3600;
    const minutes = Math.floor(seconds / 60);
    seconds = Math.floor(seconds % 60);
    let parts = [];
    if (days > 0) parts.push(`${days} day${days !== 1 ? 's' : ''}`);
    if (hours > 0) parts.push(`${hours} hour${hours !== 1 ? 's' : ''}`);
    if (minutes > 0) parts.push(`${minutes} minute${minutes !== 1 ? 's' : ''}`);
    if (seconds > 0) parts.push(`${seconds} second${seconds !== 1 ? 's' : ''}`);
    return parts.join(', ');
};