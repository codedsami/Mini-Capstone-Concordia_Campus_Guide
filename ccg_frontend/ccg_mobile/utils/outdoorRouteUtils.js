

/**
 * Calculates the distance (in meters) between two coordinates using the Haversine formula.
 */
function calculateDistance(lat1, lon1, lat2, lon2) {
    const R = 6371000; // radius of the Earth in meters
    const toRad = (deg) => (deg * Math.PI) / 180;
    const dLat = toRad(lat2 - lat1);
    const dLon = toRad(lon2 - lon1);
    const a =
        Math.sin(dLat / 2) ** 2 +
        Math.cos(toRad(lat1)) *
        Math.cos(toRad(lat2)) *
        Math.sin(dLon / 2) ** 2;
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
}

/**
 * Returns the cumulative distance along the route up to the coordinate
 * that is closest to the user's current location.
 */
export const getCumulativeDistances = (coords) => {

    const cumulative = [0];
    let total = 0;
    for (let i = 1; i < coords.length; i++) {
        const d = calculateDistance(
            coords[i - 1][1],
            coords[i - 1][0],
            coords[i][1],
            coords[i][0]
        );
        total += d;
        cumulative.push(total);
    }
    return cumulative;

}

/**
 * Finds the cumulative distance along the route corresponding to the user's current location.
 *
 * @param {Array} coords - Flattened array of coordinates.
 * @param {[number, number]} userLocation - User location in [lng, lat] format.
 * @returns {number} The cumulative distance (in meters) up to the nearest coordinate.
 */
export const getCumulativeDistanceToNearestCoord = (coords, userLocation) => {
    const cumulativeDistances = getCumulativeDistances(coords);
    let minIndex = 0;
    let minDistance = Infinity;
    for (let i = 0; i < coords.length; i++) {
        const d = calculateDistance(
            coords[i][1],
            coords[i][0],
            userLocation[1],
            userLocation[0]
        );
        if (d < minDistance) {
            minDistance = d;
            minIndex = i;
        }
    }
    return cumulativeDistances[minIndex];
}

/**
 * Returns an object with the current instruction and the remaining distance (in meters)
 * for the current step based on the cumulative distance traveled.
 */
export const getCurrentStepData = (routeData, currentDistance) => {
    let accumulatedDistance = 0;
    const steps = routeData.steps;

    for (let i = 0; i < steps.length; i++) {
        const step = steps[i];
        accumulatedDistance += step.distance;
        if (currentDistance <= accumulatedDistance) {
            const distanceIntoStep = currentDistance - (accumulatedDistance - step.distance);
            const distanceRemaining = step.distance - distanceIntoStep;
            return { instruction: step.instruction, distance: distanceRemaining };
        }
    }

    return { instruction: "Arrived at your destination", distance: 0 };
}

/**
 * Flattens the route's steps into a single array of coordinates.
 *
 * @param {Object} routeData - The route data object with a "steps" array.
 * @returns {Array} Flattened array of coordinates.
 */
export const flattenRouteCoordinates = (routeData) => {
    let allCoords = [];
    routeData.steps.forEach((step) => {
        allCoords = allCoords.concat(step.coordinates);
    });
    return allCoords;
}