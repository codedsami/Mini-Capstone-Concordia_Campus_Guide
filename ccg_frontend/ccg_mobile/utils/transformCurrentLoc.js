const transformLocation = (currentLocation) => {
    if (!currentLocation?.coords) {
        console.error("Invalid location object:", currentLocation);
        return null;
    }

    return {
        building_code: "CURR_LOC",
        civic_address: "Your Location",
        id: 1, // Placeholder ID
        location: {
            latitude: currentLocation.coords.latitude,
            longitude: currentLocation.coords.longitude,
        },
        name: "Current location",
    };
};

export default transformLocation;