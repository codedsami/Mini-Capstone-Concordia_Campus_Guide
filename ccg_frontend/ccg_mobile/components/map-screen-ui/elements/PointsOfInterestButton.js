import React from "react";
import { TouchableOpacity, Text, StyleSheet, Image } from "react-native";
import { getCategoryIcon } from "../../../utils/categoryIcons.js";

const PointsOfInterestButton = ({ type, name, isSelected, onPress }) => {
  return (
    <TouchableOpacity
      style={[styles.button, isSelected && styles.selectedButton]}
      activeOpacity={0.7}
      onPress={onPress}
      testID="points-of-interest-button"
    >
      <Image source={getCategoryIcon(type)} style={styles.icon} />
      <Text style={[styles.text, isSelected && styles.selectedText]}>{name}</Text>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "#FFFFFF", // Button color
    paddingVertical: 12,
    paddingHorizontal: 15,
    borderRadius: 30, //rounded
    marginRight: 10, // Space between buttons
    elevation: 5, // Shadow effect
  },
  selectedButton: {
    backgroundColor: "#8B1D3B", // Burgundy background when selected
  },
  text: {
    fontSize: 14,
    fontWeight: "500",
    color: "#000", // White text for contrast
  },
  selectedText: {
    color: "#FFFFFF", // White text when selected
  },
  icon: {
    width: 16,
    height: 16,
    marginRight: 5,
  },
});

export default PointsOfInterestButton;
