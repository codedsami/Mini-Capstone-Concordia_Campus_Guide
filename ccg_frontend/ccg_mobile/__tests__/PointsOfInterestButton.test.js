import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import PointsOfInterestButton from "../components/map-screen-ui/elements/PointsOfInterestButton";
import { getCategoryIcon } from "../utils/categoryIcons";

jest.mock("../utils/categoryIcons", () => ({
  getCategoryIcon: jest.fn(() => "test-file-stub"),
}));

describe("PointsOfInterestButton", () => {
  const mockOnPress = jest.fn();

  it("renders correctly with given props", () => {
    const { getByText, toJSON } = render(
      <PointsOfInterestButton type="cafe" name="Coffee" isSelected={false} onPress={mockOnPress} />
    );

    expect(getByText("Coffee")).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("applies selected styles when `isSelected` is true", () => {
    const { getByTestId } = render(
      <PointsOfInterestButton type="cafe" name="Coffee" isSelected={true} onPress={mockOnPress} />
    );

    expect(getByTestId("points-of-interest-button").props.style).toEqual(
      expect.objectContaining({
        backgroundColor: "#8B1D3B", // Burgundy background when selected
      })
    );
  });

  it("calls `onPress` when pressed", () => {
    const { getByText } = render(
      <PointsOfInterestButton type="cafe" name="Coffee" isSelected={false} onPress={mockOnPress} />
    );

    fireEvent.press(getByText("Coffee"));
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it("displays the correct icon based on `type`", () => {
    render(<PointsOfInterestButton type="cafe" name="Coffee" isSelected={false} onPress={mockOnPress} />);
    expect(getCategoryIcon).toHaveBeenCalledWith("cafe");
  });

  it("renders with default props", () => {
    const { getByText } = render(<PointsOfInterestButton type="cafe" name="Coffee" onPress={mockOnPress} />);

    expect(getByText("Coffee")).toBeTruthy();
  });

  it("renders without crashing when `onPress` is not provided", () => {
    const { getByText } = render(<PointsOfInterestButton type="cafe" name="Coffee" isSelected={false} />);

    expect(getByText("Coffee")).toBeTruthy();
  });
});
