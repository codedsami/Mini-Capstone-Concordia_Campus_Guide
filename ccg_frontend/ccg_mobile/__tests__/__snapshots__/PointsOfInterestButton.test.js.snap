// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointsOfInterestButton renders correctly with given props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderRadius": 30,
      "elevation": 5,
      "flexDirection": "row",
      "marginRight": 10,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 12,
    }
  }
  testID="points-of-interest-button"
>
  <Image
    source="test-file-stub"
    style={
      {
        "height": 16,
        "marginRight": 5,
        "width": 16,
      }
    }
  />
  <Text
    style={
      [
        {
          "color": "#000",
          "fontSize": 14,
          "fontWeight": "500",
        },
        false,
      ]
    }
  >
    Coffee
  </Text>
</View>
`;
